File: zenscript-language-server/src/main/java/raylras/zen/lsp/Semantics.java
package raylras.zen.lsp;

import org.eclipse.lsp4j.SemanticTokenModifiers;
import org.eclipse.lsp4j.SemanticTokenTypes;
import org.eclipse.lsp4j.SemanticTokensLegend;
import raylras.zen.model.symbol.Symbol.Modifier;

import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

public class Semantics {

    public static final SemanticTokensLegend SEMANTIC_TOKENS_LEGEND;

    static {
        List<String> tokenTypes = Arrays.stream(TokenType.values()).map(tokenType -> tokenType.name).collect(Collectors.toList());
        List<String> tokenModifiers = Arrays.stream(TokenModifier.values()).map(tokenModifier -> tokenModifier.name).collect(Collectors.toList());
        SEMANTIC_TOKENS_LEGEND = new SemanticTokensLegend(tokenTypes, tokenModifiers);
    }

    public static int getTokenModifiers(Modifier modifier) {
        int tokenModifiers = 0;
        switch (modifier) {
            case STATIC:
            case GLOBAL:
                tokenModifiers |= TokenModifier.STATIC.flag;
            case VAL:
                tokenModifiers |= TokenModifier.READONLY.flag;
            case VAR:
            case NONE:
                tokenModifiers |= TokenModifier.DEFINITION.flag;
        }
        return tokenModifiers;
    }

    public enum TokenType {
        TYPE(SemanticTokenTypes.Type),
        CLASS(SemanticTokenTypes.Class),
        PARAMETER(SemanticTokenTypes.Parameter),
        VARIABLE(SemanticTokenTypes.Variable),
        FUNCTION(SemanticTokenTypes.Function),
        KEYWORD(SemanticTokenTypes.Keyword),
        MODIFIER(SemanticTokenTypes.Modifier),
        COMMENT(SemanticTokenTypes.Comment),
        STRING(SemanticTokenTypes.String),
        NUMBER(SemanticTokenTypes.Number),
        OPERATOR(SemanticTokenTypes.Operator);

        public final String name;

        TokenType(String name) {
            this.name = name;
        }
    }

    public enum TokenModifier {
        DEFINITION(1, SemanticTokenModifiers.Definition),
        READONLY(2, SemanticTokenModifiers.Readonly),
        STATIC(4, SemanticTokenModifiers.Static);

        public final int flag;
        public final String name;

        TokenModifier(int flag, String name) {
            this.flag = flag;
            this.name = name;
        }
    }

}

File: zenscript-language-server/src/main/kotlin/raylras/zen/lsp/Semantics.kt
package raylras.zen.lsp

import org.eclipse.lsp4j.SemanticTokenModifiers
import org.eclipse.lsp4j.SemanticTokenTypes
import org.eclipse.lsp4j.SemanticTokensLegend
import raylras.zen.model.symbol.*

val SEMANTIC_TOKENS_LEGEND = SemanticTokensLegend().apply {
    tokenTypes = TokenType.entries.map { it.tokenName }
    tokenModifiers = TokenModifier.entries.map { it.modifierName }
}

enum class TokenType(val tokenName: String) {
    TYPE(SemanticTokenTypes.Type),
    CLASS(SemanticTokenTypes.Class),
    PARAMETER(SemanticTokenTypes.Parameter),
    VARIABLE(SemanticTokenTypes.Variable),
    FUNCTION(SemanticTokenTypes.Function),
    KEYWORD(SemanticTokenTypes.Keyword),
    MODIFIER(SemanticTokenTypes.Modifier),
    COMMENT(SemanticTokenTypes.Comment),
    STRING(SemanticTokenTypes.String),
    NUMBER(SemanticTokenTypes.Number),
    OPERATOR(SemanticTokenTypes.Operator)
}

enum class TokenModifier(val bitflag: Int, val modifierName: String) {
    DEFINITION(0b1, SemanticTokenModifiers.Definition),
    READONLY(0b10, SemanticTokenModifiers.Readonly),
    STATIC(0b100, SemanticTokenModifiers.Static)
}

val Modifiable.tokenModifier: Int
    get() = when {
        this.isGlobal || this.isStatic -> {
            TokenModifier.STATIC.bitflag or TokenModifier.READONLY.bitflag
        }

        this.isReadonly -> {
            TokenModifier.READONLY.bitflag
        }

        else -> {
            0
        }
    }

val Symbol.tokenType: TokenType?
    get() = when (this) {
        is VariableSymbol -> {
            TokenType.VARIABLE
        }

        is ParameterSymbol -> {
            // Workaround: in VSCode, it seems that read-only Parameter are not highlighted as expected, but Variable working fine.
            TokenType.VARIABLE
        }

        is ClassSymbol -> {
            TokenType.CLASS
        }

        is FunctionSymbol -> {
            TokenType.FUNCTION
        }

        is ConstructorSymbol -> {
            TokenType.FUNCTION
        }

        is OperatorFunctionSymbol -> {
            TokenType.FUNCTION
        }

        is ExpandFunctionSymbol -> {
            TokenType.FUNCTION
        }

        else -> {
            null
        }
    }
